In [ ]:
# Previous notebook execution 
%run "./1.2 Analytical View of a Dataset.ipynb"

## 1.3 Feature Screening
### Feature Screening কী?
EDA এর একটি গুরুত্বপূর্ণ ধাপ হলো Feature Screening!
এই ধাপে Dataset এর বিভিন্ন Feature বা Column পরীক্ষা করা হয় যাতে Data এর ভিতরের সমস্যা শনাক্ত করা যায়।

সহজভাবে বললে, Feature Screening হলো Dataset এর Quality Check করার প্রক্রিয়া।
এখানে সাধারণত নিচের বিষয়গুলো পরীক্ষা করা হয়:
- Missing Value আছে কি না
- Duplicate Data আছে কি না
- Irrelevant Feature আছে কি না
- Data Consistency ঠিক আছে কি না

### কেন Feature Screening প্রয়োজন?
যদি Dataset এর ভিতরে সমস্যা থাকে, তাহলে Machine Learning Model ভুল Prediction দিতে পারে।

উদাহরণ:
- Missing Value থাকলে Model Confuse হতে পারে
- Duplicate Data থাকলে Bias তৈরি হতে পারে
- Irrelevant Feature থাকলে Model Performance কমে যেতে পারে
তাই Model তৈরির আগে Feature Screening করা অত্যন্ত গুরুত্বপূর্ণ।

### Missing Value Check করা
প্রথমে Dataset এ Missing Value আছে কি না তা পরীক্ষা করা হয়।

In [17]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

### isnull() কী করে?
isnull() প্রতিটি Cell পরীক্ষা করে দেখে সেখানে Missing Value আছে কি না।
যদি Missing থাকে:
- True রিটার্ন করে

না থাকলে:
- False রিটার্ন করে

### sum() কেন ব্যবহার করা হয়?
কারণ True কে Python এ 1 হিসেবে ধরা হয়।\
তাই sum() ব্যবহার করলে প্রতিটি Column এ মোট কতগুলো Missing Value আছে তা পাওয়া যায়।

এখানে দেখা যাচ্ছে:
- Age Column এ 177 টি Missing Value আছে
- Cabin Column এ 687 টি Missing Value আছে
- Embarked Column এ 2 টি Missing Value আছে

### Missing Value কেন সমস্যা?
Missing Value থাকার কারণে:
- Statistical Calculation ভুল হতে পারে
- Visualization বিভ্রান্তিকর হতে পারে
- Machine Learning Algorithm Error দিতে পারে
- Prediction Accuracy কমে যেতে পারে

### Missing Value Handle করার কিছু উপায়
সাধারণত Missing Value Handle করার জন্য:
- Row Delete করা হয়
- Column Delete করা হয়
- Mean/Median/Mode দিয়ে Fill করা হয়
- Forward Fill বা Backward Fill ব্যবহার করা হয়

তবে কোন Method ব্যবহার করা হবে তা Dataset এর উপর নির্ভর করে।

### Duplicate Data Check করা
এখন Dataset এ Duplicate Row আছে কি না তা পরীক্ষা করা হবে।

In [10]:
df.duplicated().sum()

np.int64(0)

### duplicated() কী করে?
duplicated() প্রতিটি Row পরীক্ষা করে দেখে সেটি আগে থেকে Dataset এ আছে কি না।\
যদি Duplicate হয়:
- True দেয়

না হলে:
- False দেয়

### sum() কীভাবে কাজ করছে?
এখানেও True কে 1 হিসেবে ধরা হয়।\
তাই sum() ব্যবহার করলে মোট Duplicate Row সংখ্যা পাওয়া যায়।

অর্থাৎ Dataset এ 107 টি Duplicate Row আছে।

### Duplicate Data কেন সমস্যা?
Duplicate Data থাকার কারণে:
- Model একই তথ্য বারবার শিখে ফেলে
- Bias তৈরি হয়
- Statistical Result ভুল হতে পারে
- Dataset এর বাস্তব Distribution নষ্ট হতে পারে

### Duplicate Row Remove করা
Duplicate Data মুছে ফেলার জন্য drop_duplicates() ব্যবহার করা হয়।

In [11]:
df.drop_duplicates(inplace=True)

### drop_duplicates() কী করে?
এই Method Dataset থেকে Duplicate Row Remove করে দেয়।
### inplace=True কেন ব্যবহার করা হয়?
inplace=True ব্যবহার করলে Original Dataset এর মধ্যেই পরিবর্তন হয়ে যায়।\
অর্থাৎ:
- নতুন Dataset তৈরি হয় না
- df সরাসরি Update হয়ে যায়

### Duplicate Remove করার পর আবার Check করা
Duplicate Remove হয়েছে কি না তা আবার পরীক্ষা করা যেতে পারে।

In [12]:
df.duplicated().sum()

np.int64(0)

অর্থাৎ এখন আর কোনো Duplicate Row নেই।

### Feature Screening এ আর কী কী দেখা হয়?
বাস্তব প্রজেক্টে Feature Screening এর সময় আরও কিছু বিষয় দেখা হয়:
- Incorrect Data Type
- Constant Feature
- High Cardinality
- Irrelevant Feature
- Unique Identifier
- Data Imbalance

### Titanic Dataset এর ক্ষেত্রে পর্যবেক্ষণ
Titanic Dataset এ সাধারণত:
- Cabin এ অনেক Missing Value থাকে
- Age এ কিছু Missing Value থাকে
- Duplicate Row থাকতে পারে
- Passengerld অনেক সময় Prediction এর জন্য গুরুত্বপূর্ণ নয়

এসব বিষয় Feature Screening এর মাধ্যমে শনাক্ত করা যায়।
### Feature Screening এর মূল উদ্দেশ্য
এই ধাপের প্রধান লক্ষ্য হলো:
- Dataset Clean করা
- Problematic Feature শনাক্ত করা
- Better Analysis এর জন্য Data প্রস্তুত করা

### সারাংশ
Feature Screening হলো Dataset এর Feature বা Column গুলো পরীক্ষা করার প্রক্রিয়া।\
এই ধাপে আমরা:
- Missing Value Check করি
- Duplicate Data শনাক্ত করি
- Duplicate Remove করি
- Dataset এর Quality যাচাই করি